cmd/registry/patch/artifact.go

```diff
@@ -159,3 +205,66 @@ func applyArtifactPatch(ctx context.Context, client connection.Client, content *
 	}
 	return err
 }
+
+// kindForMimeType returns the message name to be used as the "kind" of the artifact.
+func kindForMimeType(mimeType string) string {
+	parts := strings.Split(mimeType, ".")
+	return parts[len(parts)-1]
+}
+
+// protoMessageForMimeType returns an instance of the message that represents the specified type.
+func protoMessageForMimeType(mimeType string) (proto.Message, error) {
+	messageType := strings.TrimPrefix(mimeType, "application/octet-stream;type=")
+	for k, v := range artifactMessageTypes {
+		if k == messageType {
+			return v(), nil
+		}
+	}
+	return nil, fmt.Errorf("unsupported message type %s", messageType)
+}
+
+// protoMessageForMimeType returns an instance of the message that represents the specified kind.
+func protoMessageForKind(kind string) (proto.Message, error) {
+	for k, v := range artifactMessageTypes {
+		if strings.HasSuffix(k, "."+kind) {
+			return v(), nil
+		}
+	}
+	return nil, fmt.Errorf("unsupported kind %s", kind)
+}
+
+// messageTypeForKind finds the full message type associated with a kind (typically a message name).
+func messageTypeForKind(kind string) string {
+	for k := range artifactMessageTypes {
+		if strings.HasSuffix(k, "."+kind) {
+			return k
+		}
+	}
+	return ""
+}
+
+// MimeTypeForKind returns the mime type that corresponds to a kind.
+func MimeTypeForKind(kind string) string {
+	return fmt.Sprintf("application/octet-stream;type=%s", messageTypeForKind(kind))
+}
+
+// messageFactory represents functions that construct message structs.
+type messageFactory func() proto.Message
+
+// artifactMessageTypes is the single source of truth for artifact types that can be represented in YAML.
+var artifactMessageTypes map[string]messageFactory
+
+func init() {
```

Comment:
Is this `init` function necessary? Why can't we just declare the map in the var?

Reply:
Done! I noticed that it wasn't possible to declare this as a const but didn't try the package-level var.